### PR TITLE
Refine automatic mixed precision support via hyper param

### PR DESCRIPTION
**Pull request by vinhngx**
In continuation of  https://github.com/tensorflow/tensor2tensor/pull/1637 and in response to @afrozenator 's comments in https://github.com/tensorflow/tensor2tensor/pull/1680

In this PR, we re-organize automatic mixed precision training support to provide a cleaner implementation and an easier interface via using hyper parameters. 

In particular, [GPU automatic mixed precision training](https://medium.com/tensorflow/automatic-mixed-precision-in-tensorflow-for-faster-ai-training-on-nvidia-gpus-6033234b2540) can now be enabled via setting a flag (and correspondingly a so-named hyper-parameter) `gpu_automatic_mixed_precision` for **all tensor2tensor models**, for example: 

**Transformer**
```
PROBLEM=translate_ende_wmt32k
MODEL=transformer
HPARAMS=transformer_big
DATA_DIR=/data/translate_ende_wmt32k
TRAIN_DIR=/tmp/$MODEL-$HPARAMS

t2t-trainer \
  --data_dir=$DATA_DIR \
  --problem=$PROBLEM \
  --model=$MODEL \
  --hparams_set=$HPARAMS \
  --output_dir=$TRAIN_DIR \
  --train_steps=100000 \
  --eval_steps=1000 \
  --gpu_automatic_mixed_precision=True
```

**Resnet:**
```
PROBLEM=image_imagenet224
MODEL=resnet
HPARAMS=resnet_50
DATA_DIR=/data/ImageNet
TRAIN_DIR=/tmp/$HPARAMS

t2t-trainer \
  --data_dir=$DATA_DIR \
  --problem=$PROBLEM \
  --model=$MODEL \
  --hparams_set=$HPARAMS \
  --output_dir=$TRAIN_DIR \
  --hparams='batch_size=256' \
  --worker_gpu=8 \
  --gpu_automatic_mixed_precision=True
```

This is opposed to the previous approaches of setting the OS flag `TF_ENABLE_AUTO_MIXED_PRECISION` which is a non-programatic approach, or passing the flag `gpu_auto_mixed_precision` directly to the optimizer (which will require modification of individual models to make call to optimizer with mixed precision training option).

### CODE REVIEW

File: tensor2tensor/utils/optimize.py [hunk 1]
```diff
@@ -71,8 +69,7 @@ def optimize(loss,
   opt = ConditionalOptimizer(hparams.optimizer, learning_rate, hparams, use_tpu)
   if use_tpu:
     opt = tf.contrib.tpu.CrossShardOptimizer(opt)
-  if gpu_auto_mixed_precision or os.environ.get(
-      "TF_ENABLE_AUTO_MIXED_PRECISION", "0") == "1":
+  if hparams.gpu_automatic_mixed_precision:
```
[the remark this answers is not in the document]
Comment (vinhngx) [reply tensor2tensor/utils/optimize.py]: good one. I fixed this

File: tensor2tensor/layers/common_hparams.py [hunk 2]
```diff
@@ -347,7 +347,9 @@ def basic_params1():
       num_area_layers=0,
       max_area_width=1,
       max_area_height=1,
-      memory_height=1
+      memory_height=1,
+      # Whether to use GPU automatic mixed precision (via graph rewrite)
+      gpu_automatic_mixed_precision=False
```
[the remark this answers is not in the document]
Comment (vinhngx) [reply tensor2tensor/layers/common_hparams.py]: Just add the flag again to trainer and turn hparams on accordingly.

[the remark this answers is not in the document]
Comment (vinhngx) [reply tensor2tensor/layers/common_hparams.py]: done

**Comment by vinhngx** [comment]
Thanks for the feedbacks @afrozenator . Let me know if the latest revision works.

**Comment by afrozenator** [comment]
Thanks a lot @vinhngx for contributing this in the first place and now making it better!

Will merge it in shortly.

**Comment by vinhngx** [comment]
great thanks. I'm closing #1680 then.

**Pull request merged**